[PowerRename] Fluent UX

stefansjfw opened the pull request:
## Summary of the Pull Request

**What is this about:**
New, WinUI (XAML Islands) based UI for PowerRename:
- Fresh look and feel and dark mode support! (#888)
- "Cheatsheet" flyouts provide RegEx and file datetime parameters guidance and can be added with a single click!
- Changes can now be applied, without closing the app like before. Apply-and-close is still possible by using the SplitButton. (#11759, #628)
- PowerRename settings can now be launched in a single click!
- Accessibility improvements. (#12869)

![PowerRenamevNext](https://user-images.githubusercontent.com/9866362/136649031-46b4bceb-8814-4abc-8d12-85576fc4c418.gif)

Dark mode:
![image](https://user-images.githubusercontent.com/9866362/136649051-94c78a73-7204-4d48-8715-910eaa18c257.png)


Known issues:
- [ ] Performance and memory usage when many files/folders selected

## Quality Checklist

- [X] **Linked issue:** #11759 #12869 #11564 #7101 #7097 #7081 #7036 #7025 #4129 #628 #888
- [X] **Communication:** I've discussed this with core contributors in the issue. 
- [X] **Tests:** Added/updated and all pass
- [X] **Localization:** All end user facing strings can be localized
- [X] **Installer:** Added/updated and all pass
- [ ] **Docs:** Added/ updated
- [X] **Binaries:** Any new files are added to WXS / YML
   - [ ] No new binaries
   - [X] [YML for signing](https://github.com/microsoft/PowerToys/blob/master/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [X] [WXS for installer](https://github.com/microsoft/PowerToys/blob/master/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

stefansjfw committed:
PowerRename new UI

stefansjfw commented:
> @stefansjfw do you want me to fix any of the know issues by PRing against this branch or commit directly?

Whatever you prefer :) you can commit directly, I'll check any new commit

htcfreek commented:
We should make window width smaller. On small monitors with dpi scaling this can be to big.

Commit:
Add scrollviewer

stefansjfw committed:
Don't deploy PowerRenameUI_new

Commit:
Visual updates

Commit:
Visual updates

Commit:
Updates

Commit:
Update Resources.resw

Commit:
Added docs button

Commit:
Update MainWindow.xaml

niels9001 commented:
@stefansjfw One thing we forgot is to add a link to the documentation. I have added a ? button next to the settings button, but it should navigate to: https://aka.ms/PowerToysOverview_PowerRename. Could you have a look, I don't want to touch C++ 😆

stefansjfw committed:
Wire Docs button

SeraphimaZykova reviewed (changes requested):
Great work! Only a couple of things still need to be polished. I think the filter can cause many misunderstandings and issues.

davidegiacometti commented:
I was starting to look at https://github.com/microsoft/PowerToys/issues/628 on top of this branch and I figured out that the new UI already include the requested feature.

![image](https://user-images.githubusercontent.com/25966642/137018904-cac4109b-e651-448a-9782-067b12cc0cf0.png)

With the new UI a switch in settings to keep PowerRename opened will be pointless.

niels9001 commented:
> I was starting to look at https://github.com/microsoft/PowerToys/issues/628 on top of this branch and I figured out that the new UI already include the requested feature.
> 
> ![image](https://user-images.githubusercontent.com/25966642/137018904-cac4109b-e651-448a-9782-067b12cc0cf0.png)
> 
> With the new UI a switch in settings to keep PowerRename opened will be pointless.

Thanks, added that one to the linked issues!

stefansjfw committed:
RegEx -> regular expressions

stefansjfw committed:
Update Show only renamed list on search/replace text changed

stefansjfw committed:
Update Show only renamed list on search/replace text changed - proper fix
Set searchTerm to NULL when cleared - fix Show only renamed files on clear searchTerm

stefansjfw committed:
Files/folders input error handling

stefansjfw committed:
Fix renaming with keeping UI window opened

After renaming folder, all of it's children need path update.
Without path update, further renaming of children items would
fail.

stefansjfw committed:
Update only children, not all items with greater depth

jaimecbernardo reviewed (changes requested):
I'm getting the following errors when trying to build from the "Developer Command Prompt for VS 2019":

```
C:\nuget\nuget.exe restore
MSBuild -m .\PowerToys.sln /p:Configuration=Release /p:Platform="x64"
```


```
    86>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\NuGet\16.0\Microsoft.NuGet.targe
       ts(198,5): error : Your project does not reference ".NETFramework,Version=v4.7.2" framework. Add a reference to
       ".NETFramework,Version=v4.7.2" in the "TargetFrameworks" property of your project file and then re-run NuGet res
       tore. [D:\prog\janeasystems\PowerToys\src\modules\powerrename\PowerRenameUIHost\PowerRenameUIHost.vcxproj]
```

```
Build FAILED.

       "D:\prog\janeasystems\PowerToys\PowerToys.sln" (default target) (1) ->
       "D:\prog\janeasystems\PowerToys\src\modules\powerrename\PowerRenameUIHost\PowerRenameUIHost.vcxproj.metaproj" (d
       efault target) (20) ->
       "D:\prog\janeasystems\PowerToys\src\modules\powerrename\PowerRenameUIHost\PowerRenameUIHost.vcxproj" (default ta
       rget) (86) ->
       (ResolveNuGetPackageAssets target) ->
         C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\NuGet\16.0\Microsoft.NuGet.tar
       gets(198,5): error : Your project does not reference ".NETFramework,Version=v4.7.2" framework. Add a reference t
       o ".NETFramework,Version=v4.7.2" in the "TargetFrameworks" property of your project file and then re-run NuGet r
       estore. [D:\prog\janeasystems\PowerToys\src\modules\powerrename\PowerRenameUIHost\PowerRenameUIHost.vcxproj]

    0 Warning(s)
    1 Error(s)
```

This also occurs when building for Release on Visual Studio 2019.

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		Your project does not reference ".NETFramework,Version=v4.7.2" framework. Add a reference to ".NETFramework,Version=v4.7.2" in the "TargetFrameworks" property of your project file and then re-run NuGet restore.	PowerRenameUIHost	C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\NuGet\16.0\Microsoft.NuGet.targets	198	
```

Why is .net framework 4.7.2 a requirement here?

stefansjfw committed:
Fix dictionary false positives

stefansjfw committed:
Remove .NET dep

davidegiacometti commented:
> It shouldn't be. I experienced this issue in the past, but it just stopped appearing. I also use dev cmd for VS to build PT and didn't hit this for a while. I'll try to figure out where it comes from
> 
> EDIT: @jaimecbernardo I removed .NETFramewrk dependency from project file. Let's see if this issue happens again at some point.
@jaimecbernardo not really releated to the PR but consider that there is a project that still target .NET 4.7.2
https://github.com/microsoft/PowerToys/blob/6269aa63980c43ea47fe18965c64ef00a53f5ebc/src/common/interop/PowerToysInterop.vcxproj#L12

jaimecbernardo reviewed (approved):
LGTM!
Great work!
Let's try to get a couple more reviews in.

stefansjfw committed:
Rename PowerRenameUI_new to PowerRenameUILib
Rename executable PowerRenameUIHost to PowerRename

SeraphimaZykova reviewed (approved):
Great work!

stefansjfw committed:
Merge remote-tracking branch 'origin/master' into stefan/powerrename_xaml_ui
